src/base/emu-i386/cpu.c

```diff
@@ -336,6 +330,58 @@ void raise_fpu_irq(void)
   pic_request(13);
 }
 
+int fpu_fpe_handler (unsigned char *csp)
+{
+  if (fpu_ignne) {
+    /* do some basic emulation:
+       we only deal with control FPU instructions,
+       non-control FPU instructions always cause an exception
+    */
+    if (csp[0] == 0x9b) csp++; // wait
+    if (csp[0] >= 0xd8 && csp[0] <= 0xdf) {
+      int exop = (csp[1] & 0x38) | (csp[0] & 7); // same as interp.c
+      if (exop == 0x63 || exop == 0x67)
+	exop |= ((csp[1] & 7) << 8);
+      if ((csp[1]&0xc0)==0xc0) exop |= 0x40;
+      switch (exop) {
+      case 0x31: //fstenv
+      case 0x35: //fsave
+      case 0x0263: //fclex
+      case 0x0363: //finit
+	fpu_ignne = 0;
+	g_printf("FPU_IGNNE set to %x\n", fpu_ignne);
+	/* fall through */
+      case 0x39: //fstcw
+      case 0x3d: //fstsw
+      case 0x0063: //fneni
+      case 0x0163: //fndisi
+      case 0x0463: //fnsetpm
+      case 0x0067: //fstsw %ax
+	assert(csp[-1] == 0x9b);
+	dbug_printf("coprocessor exception, skipping WAIT because of IGNNE#\n");
+	return 1;
+      case 0x21: //fldenv
+      case 0x25: //frstor
+	fpu_ignne = 0;
+	g_printf("FPU_IGNNE set to %x\n", fpu_ignne);
+	/* fall through */
+      case 0x29: //fldcw
+	dbug_printf("coprocessor exception, disabling CW load exception because of IGNNE#\n");
+	vm86_fpu_state.cwd = 0x37f;
```

Comment:
Because of this, I think
you need to leave the
"mask restore" part of
my patch. Maybe add
a value check in 0xf2
handler, add the value
that sets the mask and
remembers the original.
Then on writing 0, it
restore.